### DISCUSSION TITLE
Line number

### DISCUSSION THREAD

**Discussion by shinokada** [starship/starship#2032] [Q&A] [answered]
Is there a way to show line numbers?

For example, I'd like to run a command using a line number in zsh.

```
37$ !!/bigproj
ls /usr/local/src/bigproj
...
40$ ls -l !37:$
ls -l /usr/local/src/bigproj
```

**Comment by matchai** [comment, marked as answer]
Unfortunately, Starship is entirely stateless and doesn't have the ability to interface with the shell once the prompt is shown.
As Starship is today, this wouldn't be very possible to do.

**Comment by chipbuster** [comment]
I had a realization while reading this: we currently implement our prompt in both bash and zsh by overwriting `PS1` with a starship-generated string right before the prompt is drawn. This currently results in a shell prompt with no shell-interpolated sequences...but it doesn't have to. You can mix-and-match ANSI sequences and shell prompt sequences. As proof-of-concept, I offer up this monstrosity which I made in ZSH with a couple minutes of messing around:

![image](https://user-images.githubusercontent.com/4605384/103193421-0aeae480-48a2-11eb-91d5-60cfa947847f.png)

What I've done is make a function called `hey` which prints out both the results of `starship prompt` and then a custom ZSH prompt string. You can see that running `hey` at the terminal causes both the starship prompt (complete with ZSH zero-length escape sequences) and the raw ZSH prompt string to be printed.

Then, if I set `precmd` to assign the result of `hey` to the prompt, we can see that I get some abomination with both the starship prompt and the zsh prompt, *escaped by the ZSH shell*, to output.

In principle, we could add some sort of `passthrough` module to starship which would achieve the same thing. It would pass any strings given to it straight through into `PS1`, where it could then be escaped by the shell (we would, of course, need to make sure that that module does not make it into the escape sequence analyzer).

I don't know if this is a good idea: it ties the exact escape sequence being used to the underlying shell, which is something that starship (so far) has tried to escape. That being said, there's also some functionality that we will simply never achieve with starship's design, and this might help paper over some of those points.
